### PR TITLE
Filter branches that match regex

### PR DESCRIPTION
**Pull request by nmaludy**
Implementation of feature request #574 

This PR implements branch filtering based on a regex provided in the r10k.yaml configuration file for a given source. There is a new setting option for a source called `branch_filter:`. If specified, then only branches that match the regex will be deployed into environments. If `branch_filter:` is not specified, then all branches will be deployed. This makes it a backwards compatible change.

Please see the updated documentation for configuration in this PR for example configurations and regex patterns.

Comments and reviews appreciated.

### CODE REVIEW

**Comment by pietervogelaar** [comment]
Awesome.. just what I was looking for!

**Comment by martinseener** [comment]
Please try to merge it soon. We continiously have to wait for r10k to finish because we always have 3-4 more (dev) branches and want to get rid of the waiting for them

**Comment by andersonmills** [comment]
My apologies for not getting to this sooner. Thanks for the PR. Reviewing now.

**Comment by andersonmills** [comment]
@nmaludy So, since 3.0 is not getting released soon, we're still targeting the 2.5.x branch. Are you willing to rebase this? And, yes, that's going to end up being a new PR. :/ And yes, this will warrant a 2.6.0 release. :)

File: lib/r10k/source/git.rb [hunk 1]
```diff
@@ -113,7 +119,17 @@ def desired_contents
 
   def branch_names
     opts = {:prefix => @prefix, :invalid => @invalid_branches, :source => @name}
-    @cache.branches.map do |branch|
+    branches = @cache.branches
+    if !@branch_filter.nil? && !@branch_filter.empty?
+      branches = branches.select do |branch|
+        result = branch[/#{@branch_filter}/]
+        if result.nil?
+          logger.debug _("Branch %{branch} filtered out by branch_filter regex %{regex}") % {branch: branch, regex: @branch_filter}
```
Comment (andersonmills) [inline lib/r10k/source/git.rb]: At this point, I think this should be a "warn." We currently "warn" for skipping local modifications, and I think skipping anything feels like a warn.

File: lib/r10k/source/svn.rb [hunk 2]
```diff
@@ -108,6 +114,13 @@ def names_and_paths
 
     branches << [R10K::Environment::Name.new('production', opts), "#{@remote}/trunk"]
     @svn_remote.branches.each do |branch|
+      if !@branch_filter.nil? && !@branch_filter.empty?
+        result = branch[/#{@branch_filter}/]
+        if result.nil?
+          logger.debug _("Branch %{branch} filtered out by branch_filter regex %{regex}") % {branch: branch, regex: @branch_filter}
```
Comment (andersonmills) [inline lib/r10k/source/svn.rb]: Same: warn level.

File: doc/dynamic-environments/configuration.mkd [hunk 3]
```diff
@@ -453,3 +476,44 @@ This will create the following directory structure:
 |-- app1_production  # app1 data repository, production branch
 |-- app1_develop     # app1 data repository, develop branch
 ```
+
```
Comment (andersonmills) [inline doc/dynamic-environments/configuration.mkd:479]: This needs a "### Filtering branches" section title.

**Review by andersonmills — COMMENTED** [review]
Needs some minor tweaks and a rebase on 2.5.x and I'm happy. If I haven't heard from you by CST morning on 2017-11-16, I may do this myself only because I have a limited time to look at r10k, and I would love to get in your work. :)

**Comment by nmaludy** [comment]
Rebased with master, working on your comments now.

**Comment by nmaludy** [comment]
@andersonmills I've made all requested changes from your review. I appreciate the comments!

**Comment by andersonmills** [comment]
@nmaludy Sorry, but it's going to need to be rebased on the 2.5.x branch, not master. The problem is that master has moved on to 3.0.0-dev, and the next release of r10k is going to have to be 2.6.0. It might be a bit painful, and as I said, I'll do it tomorrow (CST), if I haven't heard from you. I know the changes there better, so it might be best if I do it, anyway.

And, by the way, thanks for this PR. It's great.

**Comment by andersonmills** [comment]
I created a new 2.6.0 branch and have ported this PR to that branch at #777.

**Comment by andersonmills** [comment]
When that PR goes in, I'll close this and merge 2.6.0 up to master.

**Comment by andersonmills** [comment]
@nmaludy Apologies that the final implementation didn't have quite the "power" that this one does.

**Comment by gregoryo2014** [comment]
Can someone please tell me the status of this work? From the comments above it appears that it got *nearly* all the way to being merged. We would dearly like to take advantage of this feature.

**Comment by andersonmills** [comment]
@gregoryo2014 This work got mutated a bit, as it's not a regex match now.

Here are docs - https://github.com/puppetlabs/r10k/blob/2.6.x/doc/dynamic-environments/configuration.mkd#ignore_branch_prefixes

This work went into r10k v2.6.0.

And thanks for reminding me that we needed to close this PR.

**Pull request closed without merging**